### PR TITLE
Incorporate Public Search Engine Comparison

### LINKED ISSUE #6: Compare Finesse's Score Against a Public Search Engine for Accuracy Comparison
## Summary
We aim to compare Finesse's Score with what a Bing search engine could produce. Bing search engines stands as our primary competitor, and it would be highly beneficial to ascertain whether our tool competes or even outperforms them. This evaluation would significantly enhance the accuracy assessment of Finesse.

## Tasks
- [x] Research and select a public search engine to compare with
- [x] Update the script with the public search engine
- [x] Filter of user queries to test only those containing at least one inspection.canada.ca URL. 
- [x] Test of user queries. Saving these queries in the results of the Finesse accuracy test.

## Acceptance Criteria
* Finesse's accuracy can be compared with the Bing search accuracy score in API tests and functions for markdown, CSV, and logging
* The integration should not affect the performance or functionality of existing features
* The comparison mechanism should clearly display the difference in accuracy between Finesse and the Bing search

### LINKED ISSUE #10: Finesse Response Times Varies a Lot
image

- [ ] Investigate why
- [ ] Apply the modifications

### LINKED ISSUE #11: Test User Generated Questions
## Summary
Develop a script to convert user-generated questions from a spreadsheet into JSON format in finesse-data, input the questions for testing in the Finesse Benchmark Tool, and finally, archive the data on the company's wiki page.

## Tasks
- [x] Create a script to parse the spreadsheet data into a JSON file in finesse-data.
- [x] Input the user-generated questions into the Finesse Benchmark Tool for testing.
- [x] Archive the generated data on the team wiki page for future reference.

## Acceptance Criteria
- The script successfully converts the spreadsheet data into a JSON file.
- The user-generated questions are accurately tested using the Finesse Benchmark Tool.
- The data is correctly archived on the company's wiki page.

### PR DESCRIPTION
**Pull request by elwoujdi**
# [Bing Search API](https://www.microsoft.com/en-us/bing/apis/pricing)
The Bing search is interesting as it is a search engine that stands out, and because we simply wanted to compare it to another popular search engine, using Bing search is an option. 

## Cons 
However, after examining the pricing, I discovered that only 1,000 transactions are free per month. This equals approximately 33 free requests per day. 2 requests are needed for a single testing (QnA) file. 

## Tasks
- [x] Refactored snippet of code 
- [x] Bing Global Web Search integration. 
- [x] Bing Search by filtering the results only for the website inspection.canada.ca. 
- [x] Reviewed calculaton of accuracy score it to consider multiple URLs for the same query. 
- [x] Development of a script to convert an excel into JSON files. 
- [x] Inclusion of a new data table for the highest and null scores. 
- [x] Fixed time response variations

## Closes
closes #6 
closes #11 
closes #10 

# Alternative considered

## Google API
Since Google Search Api limits results by 10 at a time and we have at least 20 files, each needing 100 results for testing, we will need **at least 200 requests** to obtain all the answers. **Google only offers 100 free requests by day**. After $5 are charged per 1000 requests, around $1 per test.

> num | integer: Number of search results to return. Valid values are integers between 1 and 10, inclusive.
> -- | --
> 
> [Documentation ](https://learn.microsoft.com/en-us/bing/search-apis/bing-web-search/quickstarts/rest/python)

### Library tested
[google-api-python-client](https://github.com/googleapis/google-api-python-client)

### Issues consulted
[Why Does The Google Search API Disallow More Than 100 Results? How Can I Get More?
](https://www.googlecloudcommunity.com/gc/Cloud-Hub/Why-Does-The-Google-Search-API-Disallow-More-Than-100-Results/m-p/511384#:~:text=The%20default%20number%20of%20results,100%20will%20produce%20an%20error.)

## Google web scrapping
Web scraping has been attempted as it allows for querying **completely free of charge**. However, Google has incorporated stringent security measures to limit the number of requests. Since Google displays only 10 results at a time and we have at least 20 files, each needing 100 results for testing, we will need at least 200 Google requests to obtain all the answers. Even with time delays, the 200 requests never succeed, leading to the machine's IP address being blocked for a while. Therefore, we must wait for a period longer than 30 minutes or use proxies or VPNs to work around the issue. Today, web scraping is complex and only feasible on a small scale. If we want to do it on a large scale, we need to use several VPNs and switch between them to make it undetectable.

### Libraries tested
[abenassi Google-Search-API](https://github.com/abenassi/Google-Search-API)
[Nv7-GitHub googlesearch](https://github.com/Nv7-GitHub/googlesearch)

### Issues consulted
[Github issue](https://github.com/abenassi/Google-Search-API/issues/91).
[How to fix python requests module 429 error for google search?](https://www.reddit.com/r/learnpython/comments/roqw9h/how_to_fix_python_requests_module_429_error_for/)
[Error 429 with simple query on google with requests python](https://stackoverflow.com/questions/56758333/error-429-with-simple-query-on-google-with-requests-python)
### Problem encountered
image

### CODE REVIEW

File: finesse/FINESSE_USAGE.md [hunk 1]
```diff
@@ -100,3 +101,49 @@ Accuracy statistical summary:
...
 This example shows how the CLI Output of the tool, analyzing search results from
 Azure Search and providing an accuracy score for Finesse.
+
+## Scripts
+
+### XLSX Converter to JSON 📄
+
+This script converts data from an Excel file (.xlsx) into JSON format. It is
+used for questions generated created by non-developers. It is more readable an
+excel than a json file.
```
Comment (RussellJimmies) [inline finesse/FINESSE_USAGE.md:111]: Excel files are easier to read than JSON files.

File: tests/test_bing_search.py [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+from dotenv import load_dotenv
+import os
+class TestBingSearch(unittest.TestCase):
+    def test_search_urls(self):
+        load_dotenv()
+        endpoint = os.getenv("BING_ENDPOINT")
+        subscription_key = os.getenv("BING_SEARCH_KEY")
+        bing_search = BingSearch(endpoint, subscription_key)
+
+        query = "Canada Food Inspection Agency"
```
Comment (RussellJimmies) [inline tests/test_bing_search.py:12]: Canadian Food Inspection Agency

**Review by RussellJimmies — CHANGES_REQUESTED** [review]
Looks good to me! Small typos

**Comment by SonOfLope** [comment]
Failing pipeline will be fixed with https://github.com/ai-cfia/github-workflows/pull/110

File: finesse/FINESSE_USAGE.md (hunk 1 above)
Comment (ibrahim-kabir) [reply finesse/FINESSE_USAGE.md:111]: Done

File: tests/test_bing_search.py (hunk 2 above)
Comment (ibrahim-kabir) [reply tests/test_bing_search.py:12]: Done

File: finesse/scripts/xlsx_converter_json.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+parser.add_argument('--file-name', dest='file_name', help='Name of the input file')
+parser.add_argument('--sheet-name', dest='sheet_name', default='To fill', help='Name of the worksheet')
+
+args = parser.parse_args()
+
+INPUT_FOLDER = args.input_folder
+OUTPUT_FOLDER = args.output_folder
+FILE_NAME = args.file_name
+SHEET_NAME = args.sheet_name
+FILE_PATH = INPUT_FOLDER+FILE_NAME
```
Comment (rngadam) [inline finesse/scripts/xlsx_converter_json.py:18]: this assume that the input folder ends with a /

better to use os.path.join to ensure platform-independant file joining.

**Review by rngadam — CHANGES_REQUESTED** [review]
will need to be rebased as it conflicts with files on the main branch (and also, we'd want to see that all checks passes. Right now, no checks are running.

File: finesse/scripts/xlsx_converter_json.py (hunk 3 above)
Comment (ibrahim-kabir) [reply finesse/scripts/xlsx_converter_json.py]: Done

Comment (rngadam) [reply finesse/scripts/xlsx_converter_json.py]: Note that reviewer should resolve the conversation, not the author.

**Review by rngadam — APPROVED** [review]

File: finesse/bing_search.py [hunk 4]
```diff
@@ -0,0 +1,42 @@
+from azure.cognitiveservices.search.websearch import WebSearchClient
+from msrest.authentication import CognitiveServicesCredentials
+import time
+import statistics
```
Comment (MaxenceGui) [inline finesse/bing_search.py:4]: Import orders 👆🏻

File: finesse/finesse_test.py [hunk 5]
```diff
@@ -2,9 +2,11 @@
 from jsonreader import JSONReader
 import os
 import json
-from accuracy_functions import save_to_markdown, save_to_csv, log_data, calculate_accuracy
+from accuracy_functions import save_to_markdown, save_to_csv, calculate_accuracy, update_dict_bing_data
 from host import is_host_up
```
Comment (MaxenceGui) [inline finesse/finesse_test.py:6]: Import orders 👆🏻

File: finesse/finesse_test.py [hunk 6]
```diff
@@ -58,39 +60,48 @@ def search_accuracy(self):
...
-            save_to_csv(self.qna_results, self.engine)
-
     def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
         self.path = self.environment.parsed_options.path
         self.engine = self.environment.parsed_options.engine
         self.format = self.environment.parsed_options.format
         self.once = self.environment.parsed_options.once
         self.top = self.environment.parsed_options.top
+        settings["engine"] = self.engine
```
Comment (MaxenceGui) [inline finesse/finesse_test.py:92]: usually `__init__` is the first function declared after the class declaration

**Review by MaxenceGui — COMMENTED** [review]
You have a lot of function that I feels could be generalize so other AI-Lab project can beneficiate from them. It could be cool to start a folder of common function that can be used for more than 1 project.

File: finesse/finesse_test.py (hunk 5 above)
Comment (ibrahim-kabir) [reply finesse/finesse_test.py:6]: fixed in #9

File: finesse/bing_search.py (hunk 4 above)
Comment (ibrahim-kabir) [reply finesse/bing_search.py:4]: fixed in #9

File: finesse/finesse_test.py (hunk 6 above)
Comment (ibrahim-kabir) [reply finesse/finesse_test.py:92]: fixed on #9

**Comment by ibrahim-kabir** [comment]
@RussellJimmies, do you have any more request changes ?

**Review by MaxenceGui — APPROVED** [review]

**Comment by ibrahim-kabir** [comment]
@RussellJimmies your change request is blocking my merge

**Review by RussellJimmies — APPROVED** [review]

**Pull request merged**